python/hsfs/constructor/query.py

```diff
@@ -520,31 +530,172 @@ def left_feature_group_start_time(self, left_feature_group_start_time):
     def left_feature_group_end_time(self, left_feature_group_end_time):
         self._left_feature_group_end_time = left_feature_group_end_time
 
-    @property
-    def features(self):
-        return self._left_features
-
     def append_feature(self, feature):
         """
-        !!! example
-            ```python
-            fg1 = fs.get_feature_group("...")
-            fg2 = fs.get_feature_group("...")
+        Append a feature to the query.
 
-            query = fg1.select_all().join(fg2.select_all())
-
-            query.append_feature('feature_name')
-            ```
+        # Arguments
+            feature: `[str, Feature]`. Name of the feature to append to the query.
         """
+        feature = util.validate_feature(feature)
+
         self._left_features.append(feature)
 
+        return self
+
     def is_time_travel(self):
+        """Query contains time travel"""
         return (
             self.left_feature_group_start_time
             or self.left_feature_group_end_time
             or any([_join.query.is_time_travel() for _join in self._joins])
         )
 
+    def is_cache_feature_group_only(self):
+        """Query contains only cached feature groups"""
+        return all(
+            [isinstance(fg, feature_group.FeatureGroup) for fg in self.featuregroups]
+        )
+
+    def _get_featuregroup_by_feature(self, feature: Feature):
+        # search for feature by id, and return the fg object
+        fg_id = feature._feature_group_id
+        for fg in self.featuregroups:
+            if fg.id == fg_id:
+                return fg
+
+        # search for feature by name and collect fg objects
+        featuregroup_features = {}
+        for feat in self._left_feature_group.features:
+            featuregroup_features[feat.name] = featuregroup_features.get(
+                feat.name, []
+            ) + [self._left_feature_group]
+        for join_obj in self.joins:
+            for feat in join_obj.query._left_feature_group.features:
+                featuregroup_features[feat.name] = featuregroup_features.get(
+                    feat.name, []
+                ) + [join_obj.query._left_feature_group]
```

Comment:
something similar could also be done for assigning `query_features` in `_get_feature_by_name`.

Reply:
It's a different logic, because the `features` attribute only contains features that were selected in the "select" part of the query. But filters can be defined on any of the columns in the underlying featuregroups, therefore flyingduck needs this list here to figure out to which featuregroup the features in the filter belong (we cannot rely on the feature.featuregroup_id because we allow this type of statement: `fg.select_all(Feature("name") > 10)` where the Feature object in not bound to a featuregroup, hence the `.featuregroup_id` is empty.